service-base/src/main/java/org/eclipse/hono/service/plan/PrometheusBasedTenantPlanValidator.java

```diff
@@ -0,0 +1,125 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.service.plan;
+
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.ext.web.client.HttpResponse;
+import io.vertx.ext.web.client.WebClient;
+import org.eclipse.hono.service.metric.MicrometerBasedMetrics;
+import org.eclipse.hono.util.TenantObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * An implementation of {@link TenantPlanValidator} which uses metrics data from the prometheus backend to check if
+ * further connections or messages are allowed by comparing with the configured limits.
+ */
+@Component
+public final class PrometheusBasedTenantPlanValidator extends TenantPlanValidator {
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+    private final WebClient client;
+    private PrometheusBasedTenantPlanValidatorProperties config;
+
+    /**
+     * Creates a new PrometheusBasedTenantPlanValidator instance.
+     *
+     * @param vertx The Vertx instance to use.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public PrometheusBasedTenantPlanValidator(final Vertx vertx) {
+        this.client = WebClient.create(Objects.requireNonNull(vertx));
+    }
+
+    @Override
+    public Future<?> isFurtherConnectionsAllowed(final TenantObject tenant) {
+        Objects.requireNonNull(tenant);
+        return checkConnectionsLimit(tenant);
+    }
+
+    /**
+     * Sets configuration properties.
+     *
+     * @param config The configuration properties.
+     * @throws NullPointerException if props is {@code null}.
+     */
+    @Autowired
+    public void setConfig(final PrometheusBasedTenantPlanValidatorProperties config) {
+        Objects.requireNonNull(config);
+        this.config = config;
+    }
+
+    private Future<?> checkConnectionsLimit(final TenantObject tenant) {
```

Comment:
IMHO this should actually be the implementation of *isFurtherConnectionsAllowed* ...

Reply:
Done